dbms/src/AggregateFunctions/AggregateFunctionHistogram.h

```diff
@@ -0,0 +1,363 @@
+        storage.reserve(2 * size - max_bins);
+        std::priority_queue<QueueItem, std::vector<QueueItem>, std::greater<QueueItem>> queue;
+
+        auto quality = [&](int i) { return points[next[i]].mean - points[i].mean; };
+        for (size_t i = 0; i + 1 < size; i++)
+            queue.push({quality(i), i});
+
+        while (newsz > max_bins && !queue.empty())
+        {
+            auto min_item = queue.top();
+            queue.pop();
+            auto l = min_item.second;
+            auto r = next[l];
+            if (!active[l] || !active[r] || quality(l) > min_item.first)
+                continue;
+
+            points[l] = points[l] + points[r];
+
+            delete_node(r);
+            if (active[next[l]])
+                queue.push({quality(l), l});
+            if (active[previous[l]])
+                queue.push({quality(previous[l]), previous[l]});
+
+            newsz--;
+        }
+
+        size_t l = 0;
+        for (size_t r = 0; r < size; r++)
+            if (active[r])
+                points[l++] = points[r];
+        size = newsz;
+    }
+
+    /***
+     * Delete too close points from histogram.
+     * Assumes that points are sorted.
+     */
+    void unique()
+    {
+        size_t l = 0;
+        for (auto r = l + 1; r < size; r++)
+        {
+            if (points[l].mean + epsilon >= points[r].mean)
+            {
+                points[l] = points[l] + points[r];
+            }
+            else
+            {
+                l++;
+                points[l] = points[r];
+            }
+        }
+        size = l + 1;
+    }
+
+    void init(Arena* arena)
+    {
+        points = reinterpret_cast<WeightedValue*>(arena->alloc(max_bins * 2 * sizeof(WeightedValue)));
+    }
+
+public:
+    AggregateFunctionHistogramData(UInt32 max_bins)
+        : max_bins(max_bins)
+        , size(0)
+        , lower_bound(std::numeric_limits<Mean>::max())
+        , upper_bound(std::numeric_limits<Mean>::min())
+        , points(nullptr)
+    {
+    }
+
+    void insertResultInto(ColumnVector<Mean>& to_lower, ColumnVector<Mean>& to_upper, ColumnVector<Weight>& to_weights) {
+        if (!points) return;
+
+        compress();
+        unique();
+
+        for (size_t i = 0; i < size; i++)
+        {
+            to_lower.insert((i == 0) ? lower_bound : (points[i].mean + points[i - 1].mean) / 2);
+            to_upper.insert((i + 1 == size) ? upper_bound : (points[i].mean + points[i + 1].mean) / 2);
+
+            // linear density approximation
+            Weight lower_weight = (i == 0) ? points[i].weight : ((points[i - 1].weight) + points[i].weight * 3) / 4;
+            Weight upper_weight = (i + 1 == size) ? points[i].weight : (points[i + 1].weight + points[i].weight * 3) / 4;
+            to_weights.insert((lower_weight + upper_weight) / 2);
+        }
+    }
+
+    void add(Mean value, Weight weight, Arena* arena)
+    {
+        if (!points)
+            init(arena);
+        points[size++] = {value, weight};
```

Comment:
We will use foreign memory if `max_bins == 0`.

Reply:
max_bins=0 is strange case. I think function should accept only positive bin count.